design-history/exploration-reports/2019.06-unixfsv2-spike-01.md

```diff
@@ -0,0 +1,296 @@
+A unixfsv2 spike, using schemas
+===============================
+
+This document is an exploration of what a hypothetical "unixfs v2" might look like,
+using schemas as a design language for describing the data structures.
+
+This is not a final proposal, but rather a lens to explore both unixfsv2
+concepts and schemas themselves at the same time, and how we expect them
+to contribute to each other.  Several concepts in schemas, advanced layouts,
+their interactions, and their syntaxes will be trialled here; they are not final.
+
+Without further ado...
+
+Schema
+------
+
+```
+type AnyFile union {
+	| File "f"
+	| Dir "d"
+	| Symlink "s"
+} representation keyed
+
+type Filename string
+
+type Dir struct {
+	members {Filename:DirEnt}<HAMT>
+}
+
+type DirEnt struct {
+	content &AnyFile
+	attribs Attribs
+}
+
+type File struct {
+	content Bytes<Rabin>
+}
+
+type Symlink struct {
+	target String
+}
+
+advanced Rabin {
+	kind bytes
+	implementation "IPLD/experimental/rabin/v1"
+}
+
+advanced HAMT {
+	kind map
+	implementation "IPLD/experimental/HAMT/v1"
+}
+
+type Attribs struct {
+	mtime Int # Danger danger will robinson; int sizes will rear their ugly head.
```

Comment:
Optionality of schema's aside, on the subject of `mtime` these days I would say subseconds are mandatory. This is how I extended the protobuf in UFS1.5:
```
// A generic high precision time specification
message Timespec {
       // Use of int64 is deliberate - negative epoch is super unlikely
       // Using sint64 would introduce zig-zag encoding ( harder to eyeball )
       // The varint representing the time of writing this is 5 bytes long
       // It will remain so until October 26, 3058 ( 34,359,738,367 )
       required int64 UnixEpoch = 1;

       // fixed32 ( always 4 bytes, no varint ), as nanosecs are often > 2^28
       // https://developers.google.com/protocol-buffers/docs/proto#scalar
       // NOTE: on the wire this value will be *little* endian
       optional fixed32 UnixEpochNsec = 2;
}
```

Similar sizing considerations will apply to CBOR. I would strongly recommend reconsidering use of 32bit ints...